### PR TITLE
[1.x] Inertia: Automatically publish & register Inertia Middleware

### PR DESCRIPTION
**Pull request by claudiodekker**
Solves #326 

This PR automatically publishes and registers the new `HandleInertiaRequests` Inertia Middleware with the Http Kernel, to enable automatic asset versioning (cache busting) and error sharing as of default. To take full advantage of this, it also enables versioning as of default in the `webpack.mix.js`.

Additionally, because Jetstream [automatically registers the `ShareInertiaData` middleware on the 'web' group as priority middleware](https://github.com/laravel/jetstream/pull/327/files#diff-232229d3c322d98dbcce6ac50983b7ec119b365097def676313c286fa15b6d00R206-R207), this PR also adds logic to ensure the `HandleInertiaRequests` middleware always runs directly after it when it exists. This gives the user the ability to overwrite the Inertia built-in defaults (such as the `user` key) if they want to.

Finally, because this Middleware will exist entirely in userland, it allows the user to change it in to any configuration they want, including disabling it entirely.

### CODE REVIEW

File: src/Http/Middleware/ShareInertiaData.php [hunk 1]
```diff
@@ -20,6 +20,10 @@ class ShareInertiaData
      */
     public function handle($request, $next)
     {
+        Inertia::version(function () {
+            return md5_file(public_path('mix-manifest.json'));
```
Comment (m1guelpf) [inline src/Http/Middleware/ShareInertiaData.php:24]: What if this file didn't exist? Wouldn't that make Laravel automatically 500?

**Review by m1guelpf — CHANGES_REQUESTED** [review]

File: src/Http/Middleware/ShareInertiaData.php (hunk 1 above)
Comment (reinink) [reply src/Http/Middleware/ShareInertiaData.php:24]: Yup.

Comment (claudiodekker) [reply src/Http/Middleware/ShareInertiaData.php:24]: Good point. I'll do some more testing on this, including seeing if I can find a good solution for Vapor.

Comment (m1guelpf) [reply src/Http/Middleware/ShareInertiaData.php:24]: For Vapor, you can use `config('app.asset_url')`, but my point is that this shouldn't be on vendor code.

Comment (claudiodekker) [reply src/Http/Middleware/ShareInertiaData.php]: Yeah, I already had a fix for it ready, I just didn't get around to testing it yet as I got interrupted by some other things. I just pushed the changes. Do you still think it's a bad idea now?

I could move it to a publishable Service Provider, but then it'd always boot. 
The nice thing about the current solution (middleware) is that it's only booted when it's actually being used.

**Comment by claudiodekker** [comment]
Alright, so, tons of changes. Let me guide you through it.

First, I've solved the issue of the `Inertia::version` being part of the vendor, by moving it to a newly-published `InertiaServiceProvider`. 

Because as a result of this we ended up with a lot of duplicated logic (`installFortifyServiceProvider`, `installJetstreamServiceProvider`, `installInertiaServiceProvider`), I've refactored this methods to a single `installServiceProvider($name, $after)` method.

Next, I ran into the dilemma that while Jetstream's `Inertia::share` would be nice in userland, doing so would cause a lot of repetition. We already had two different files (`JetstreamServiceProvider.php`, `JetstreamWithTeamsServiceProvider.php`), and making this change would result in 4 files, as we'd also need one for Inertia as well as one for Inertia with Teams.

To solve this issue, I remembered a thing I learned from @taylorotwell (Laravel ☁️ ) where he used Blade views to 'generate' scripts. I basically did the same thing by moving the `JetstreamServiceProvider` into a single 'template', and rendering the file into the state in which it's needed on-the-fly.

Hope this makes sense!

**Comment by taylorotwell** [comment]
To me it feels odd to call Inertia::share outside of a web request context. For example, I would pretty much expect it to be done in a middleware or controller where there is definitely an HTTP request happening. A console command will fire up all service providers and this will call Inertia::share in that case.

**Comment by claudiodekker** [comment]
Yeah, I definitely understand, but the inability to define your own defaults and override the user there is kind of problematic, as it'd force you to pretty much drop Jetstream in that situation.

Personally, I don't think its too big of a problem to have them bootstrap like this. Yes, it's maybe not the most efficient, but I kind of see Inertia::share as a way to register container bindings that aren't resolved yet

Perhaps there's something else we can try? Maybe even a change that we can make upstream in Inertia's Laravel adapter?

**Comment by claudiodekker** [comment]
Done. I've reverted the changes and will update the original message of this PR to reflect this as well.
(Sorry for the frequent force-pushes, having a bit of a fight here here with my machine.. )

In addition, I've briefly had a chat about this topic with @reinink, since we're using `Inertia::share` inside the `AppServiceProvider` on our [Inertia best-practices / demo repository (PingCRM)](https://github.com/inertiajs/pingcrm). I'll create a PR to that repository over the next few days to use Middleware instead, similar to what you're doing here. 👍

**Comment by claudiodekker** [comment]
I've also gone ahead and enabled `.version()` in the `webpack.mix.js` to take advantage of versioning as of default, as was suggested in #326. I do realize it's now also enabled for Livewire, but I think that's fine?

**Comment by taylorotwell** [comment]
Is this ready? If so, can you mark it as ready. I have my GitHub set to filter out draft PRs.

**Comment by claudiodekker** [comment]
Yeah, it's ready, but @reinink asked me to re-draft it so he could talk to you about it.. Still want me to mark it as ready?

File: src/JetstreamServiceProvider.php [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 namespace Laravel\Jetstream;
 
+use App\Http\Middleware\PrepareInertia;
```
Comment (bonzai) [inline src/JetstreamServiceProvider.php:5]: This will not work if my namespace is not `App`.

**Comment by taylorotwell** [comment]
Status? Should this be made active?

**Comment by claudiodekker** [comment]
Sorry, forgot to post an update here.

Because of the conversation that happened here, we've decided to change Inertia's Laravel adapter to use Middleware as-of default for sharing data etc. as well, resulting in this PR: https://github.com/inertiajs/inertia-laravel/pull/161

The plan right is to release that PR tomorrow, at which point I'll update this one and will mark it ready for review.

**Comment by claudiodekker** [comment]
Alright, so, I think this is ready, but I'll test it to make sure tomorrow.

**Comment by claudiodekker** [comment]
@taylorotwell Alright, everything seems to be working 👍  

I also went ahead and split the `webpack.mix.js` file to be stack-specific, because we already had [an ongoing discussion about this over here](https://github.com/laravel/jetstream/pull/328#issuecomment-706404285), as well as that I didn't really like the idea of touching something that Livewire also gets affected by if I don't have to.

**Comment by taylorotwell** [comment]
@claudiodekker  Can this be re-targeted to master so we can release all of these Inertia improvements at once?

**Comment by claudiodekker** [comment]
@taylorotwell Sure, give me 5 mins.. verifying that everything will keep working

**Comment by claudiodekker** [comment]
@taylorotwell Done 👍

**Comment by taylorotwell** [comment]
I have a feeling all that middleware replacement using PHP_EOL isn't going to work on Windows. 👀

File: src/Console/InstallCommand.php [hunk 3]
```diff
@@ -424,21 +411,52 @@ protected function ensureApplicationIsTeamCompatible()
...
+     *
+     * @param  string $after
+     * @param  string $name
+     * @param  string $group
+     * @return void
+     */
+    protected function installMiddlewareAfter($after, $name, $group = 'web')
+    {
+        $httpKernel = file_get_contents(app_path('Http/Kernel.php'));
+        $middlewareGroups = Str::before(Str::after($httpKernel, '$middlewareGroups = ['.PHP_EOL), '];'.PHP_EOL);
```
Comment (taylorotwell) [inline src/Console/InstallCommand.php:442]: Particularly here. On Windows I'm not sure this Str::after call will return what you want.

Comment (claudiodekker) [reply src/Console/InstallCommand.php:442]: Yeah, I think you're right.. I totally forgot about Windows doing things like this differently.

Let me see if I can adjust this.

**Comment by taylorotwell** [comment]
Honestly my gut would be to just replace the HTTP kernel entirely with a stub that contains exactly what you want. It will have some duplication with Laravel itself but would prevent the need for this complicated replace logic.

**Comment by claudiodekker** [comment]
Yeah, agreed, but it does add extra maintenance if there's a change made in Laravel itself.

**Comment by claudiodekker** [comment]
@taylorotwell How's this? Or do you still prefer creating a HTTP kernel stub?

**Comment by taylorotwell** [comment]
image

I get this. I assume because you're installing the Inertia dependency but then calling a command in the same process such that for that process it is not installed yet. Make sense?

**Comment by claudiodekker** [comment]
Ah shoot! Yeah, I get what you're saying, it's re-using the container that's already booted.

**Comment by taylorotwell** [comment]
Probably you would need to use Symfony Process to run the command.

**Comment by taylorotwell** [comment]
OK seems fine to me now. Good to merge?

**Comment by claudiodekker** [comment]
Yep, sure. I just got a pretty crazy idea that might actually be a better alternative to the 'patching' that I'm doing now, but I can just PR that later as it's not breaking 👍

**Comment by claudiodekker** [comment]
@taylorotwell In case you were curious: I played around with that idea I mentioned for a bit, but I don't think the end result would be better.

Basically, what I wanted to do, is use [ReflectionClass::getDefaultProperties](https://www.php.net/manual/en/reflectionclass.getdefaultproperties.php) to load the Kernel's class properties, and add the middleware to that. Then, I'd basically implode the array, and inject it into a `Kernel.stub` file, followed by overwriting the result to the application's `Kernel.php`

While this would work just fine, all comments or middleware that's included but 'commented' by default would be stripped out, and without a good way to grab those I think the end result would be less useful compared to what we have now.

**Pull request merged**